src/bundle.rs

```diff
@@ -10,142 +10,34 @@ use nonempty::NonEmpty;
 use zcash_note_encryption::{try_note_decryption, try_output_recovery_with_ovk};
 
 use crate::{
+    action::Action,
     address::Address,
     bundle::commitments::{hash_bundle_auth_data, hash_bundle_txid_data},
     circuit::{Instance, Proof, VerifyingKey},
     keys::{IncomingViewingKey, OutgoingViewingKey},
-    note::{ExtractedNoteCommitment, Note, Nullifier, TransmittedNoteCiphertext},
+    note::Note,
     note_encryption::OrchardDomain,
     primitives::redpallas::{self, Binding, SpendAuth},
     tree::Anchor,
     value::{ValueCommitTrapdoor, ValueCommitment, ValueSum},
 };
 
-/// An action applied to the global ledger.
-///
-/// Externally, this both creates a note (adding a commitment to the global ledger),
-/// and consumes some note created prior to this action (adding a nullifier to the
-/// global ledger).
-///
-/// Internally, this may both consume a note and create a note, or it may do only one of
-/// the two. TODO: Determine which is more efficient (circuit size vs bundle size).
-#[derive(Debug, Clone)]
-pub struct Action<A> {
-    /// The nullifier of the note being spent.
-    nf: Nullifier,
-    /// The randomized verification key for the note being spent.
-    rk: redpallas::VerificationKey<SpendAuth>,
-    /// A commitment to the new note being created.
-    cmx: ExtractedNoteCommitment,
-    /// The transmitted note ciphertext.
-    encrypted_note: TransmittedNoteCiphertext,
-    /// A commitment to the net value created or consumed by this action.
-    cv_net: ValueCommitment,
-    /// The authorization for this action.
-    authorization: A,
-}
-
 impl<T> Action<T> {
```

Comment:
Why is this still allowed to be here without violating the Orphan rule?

Reply:
The orphan rule applies to crates, not modules. We're still inside the `orchard` crate; it doesn't matter where in it we impl these. See also how we implemented `zcash_address`.